Rename alert status OK to Recovered and fix some UX issues around disabling a rule while being in an error state

mikecote opened the pull request:
Resolves #87055
Resolves #98073

In this fix-it PR, I'm fixing a couple of UX issues around the rule details page.

1. When a rule is in an error state (based on `lastExecutionStatus`), the error banner will re-display after the user disables then enables the rule if the banner was previously dismissed. This way, if ever the rule does encounter a different error after being enabled, the error gets displayed regardless if they chose to dismiss the previous error or not.
2. The rule details page no longer displays the error message when the rule is disabled when in an error state (based on `lastExecutionStatus`). A banner is already displayed to indicate the rule is disabled.
3. When enabling a rule, the execution status gets updated to `pending` just like when a rule is created. This removes the confusion of seeing an error message in the rule details page when the rule didn't finish running yet (after being re-enabled).
4. Renames alert instance status "OK" to "Recovered".

elasticmachine commented:
Pinging @elastic/kibana-alerting-services (Team:Alerting Services)

pmuellr commented:
I can't find an issue for this, but I've been thinking it would be more useful for us to show instances that have recovered with a visual status of "recovered", instead of "ok" as it does today.  Makes it more clear that it **WAS** active, previously.  And set expectations that you aren't seeing all the potential range of alert instances, only those that have "alerted".

Haven't looked at the size of this issue yet, but if we're doing a couple clean ups of this kind of stuff, this PR could be a place to put this.  I think it involves just a change in the text we display in the instance table.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/)
* Commit: 7bca356f565f34d0410a7c0f90f1eb3978fc62fe
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-98135/7bca356f565f34d0410a7c0f90f1eb3978fc62fe)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/7598f57bcb57f5a9996570a87a6c62c36b741974) | [after](https://github.com/elastic/kibana/commit/7bca356f565f34d0410a7c0f90f1eb3978fc62fe) | diff |
| --- | --- | --- | --- |
| `triggersActionsUi` | 1.5MB | 1.5MB | +86.0B |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @mikecote

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/","number":"122027","commit":"7bca356f565f34d0410a7c0f90f1eb3978fc62fe"}]}
PIPELINE-->

YulNaumenko reviewed (approved):
LGTM

mikecote commented:
@pmuellr it was pretty easy to change, I did the change within https://github.com/elastic/kibana/pull/98135/commits/c7b24857390ebeaec8cbcd53f92af5ff674cd553.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/)
* Commit: c7b24857390ebeaec8cbcd53f92af5ff674cd553
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-98135/c7b24857390ebeaec8cbcd53f92af5ff674cd553)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup10](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/execution/node/753/log)
* [Execute xpack-kibana-ciGroup10](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/execution/node/891/log)

### History
* :green_heart: [Build #122027](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/) succeeded 7bca356f565f34d0410a7c0f90f1eb3978fc62fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @mikecote

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/","number":"122027","commit":"7bca356f565f34d0410a7c0f90f1eb3978fc62fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/","number":"122126","commit":"c7b24857390ebeaec8cbcd53f92af5ff674cd553"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/)
* Commit: 6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-98135/6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup12](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/execution/node/768/log)
* [Execute xpack-kibana-ciGroup12](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/execution/node/784/log)

### History
* :broken_heart: [Build #122126](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/) failed c7b24857390ebeaec8cbcd53f92af5ff674cd553
* :green_heart: [Build #122027](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/) succeeded 7bca356f565f34d0410a7c0f90f1eb3978fc62fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @mikecote

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/","number":"122027","commit":"7bca356f565f34d0410a7c0f90f1eb3978fc62fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/","number":"122126","commit":"c7b24857390ebeaec8cbcd53f92af5ff674cd553"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/","number":"122160","commit":"6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b"}]}
PIPELINE-->

pmuellr commented:
re: ok -> recovered.  I happened to think - "do we have doc to fix?  maybe a screen shot?".  Took a brief look - nope!  Did I miss it?  It ain't right that we don't have doc on the alert details UI!  Let's open a new doc issue if we don't.

I think this is probably release-note worthy though.  It's possible some customers may be confused by the change of "OK" to "Recovered".  We should note that the semantics have not changed; these have always been alert instances which were previously active but are no longer active, we just changed the label to better reflect that state.

ymao1 commented on x-pack/plugins/alerting/server/alerts_client/alerts_client.ts:
```diff
@@ -953,6 +953,11 @@ export class AlertsClient {
         updatedAt: new Date().toISOString(),
+        executionStatus: {
+          status: 'pending',
+          lastExecutionDate: new Date().toISOString(),
```
nit: Does it make sense to reset the `lastExecutionDate` here? I guess it would quickly be overwritten the next time the rule runs, but technically, this date is not the last time the rule was executed.

pmuellr replied:
It looks like we do set that field to the same "current date" for other cases of `status: 'pending'`.  So, it's consistent.  I think we picked the name `lastExecutionDate` before starting to use `pending`, and so ... the name kinda doesn't make sense any more.  Would be better described (I think) as "lastStateUpdateDate" or something.

mikecote replied:
>Would be better described (I think) as "lastStateUpdateDate" or something.

I think that makes sense. Should I create an issue for 8.0 release? This way we can make it a breaking change.

I think copying how it's done on create is ok for now but definitely strange.

pmuellr replied:
DateDateDate

I'm not against changing the name to better reflect reality, especially as an 8.0 thing.  Though I wonder if `lastStateUpdate` makes sense - the date value is supposed to reflect the last execution, except we knew we didn't have that date in some previous 7.x release, and obviously don't have one for newly created rules.  So it does kind of make sense to continue with the current name, with the caveat that you need to check to see if the `status` is 'pending' which means "it hasn't run yet".  Basically, if we call it `lastStateUpdate`, would folks expect other changes might also change this date?  What happens to these fields when disabling a rule?

mikecote replied:
I read this as "the execution status is pending, and it's been pending since `lastExecutionDate`". So any name that can reflect that and last time the rule executed I'm +1. Seems like a separate issue to be created?

ymao1 commented on x-pack/plugins/triggers_actions_ui/public/application/sections/alert_details/components/alert_details.test.tsx:
```diff
@@ -463,6 +465,74 @@ describe('disable button', () => {
     expect(enableAlert).toHaveBeenCalledTimes(1);
   });
+
+  it('should bring back error banner after re-enable if previously dismissed', async () => {
```
Is it expected that I will see the error banner without refreshing the page after reenabling? I see the error banner disappear after disabling but after enabling, I have to refresh the page to see the error banner again

![Apr-26-2021 08-23-42](https://user-images.githubusercontent.com/13104637/116081893-cb25e500-a668-11eb-909c-a38f830225e6.gif)

mikecote replied:
>Is it expected that I will see the error banner without refreshing the page after reenabling? I see the error banner disappear after disabling but after enabling, I have to refresh the page to see the error banner again

After re-enabling the rule, it would have a `pending` status which I believe explains why the banner doesn't re-appear until you refresh the page (or click the refresh button). Because by then, the rule had a chance to run once and update the execution status. Does this experience seem strange?

ymao1 replied:
Ah I see. I think that's fine. The test name made me think the error banner should reappear right away but I think it's ok if it doesn't.

mikecote replied:
Ah, I see what you mean. I fixed the test name in this commit: https://github.com/elastic/kibana/pull/98135/commits/624ae75701b90bca54127aca0a6ce980a2047785.

ymao1 commented:
> re: ok -> recovered. I happened to think - "do we have doc to fix? maybe a screen shot?". Took a brief look - nope! Did I miss it? It ain't right that we don't have doc on the alert details UI! Let's open a new doc issue if we don't.

https://www.elastic.co/guide/en/kibana/master/rule-details.html We do have a screenshot for this

mikecote commented:
>> re: ok -> recovered. I happened to think - "do we have doc to fix? maybe a screen shot?". Took a brief look - nope! Did I miss it? It ain't right that we don't have doc on the alert details UI! Let's open a new doc issue if we don't.
>
>https://www.elastic.co/guide/en/kibana/master/rule-details.html We do have a screenshot for this

I will update those for sure!

mikecote commented:
>I think this is probably release-note worthy though. It's possible some customers may be confused by the change of "OK" to "Recovered". We should note that the semantics have not changed; these have always been alert instances which were previously active but are no longer active, we just changed the label to better reflect that state.

I went ahead and updated the PR title and release note label for this 🙏

mikecote commented on x-pack/test/alerting_api_integration/security_and_spaces/tests/alerting/enable.ts:
```diff
@@ -119,6 +119,7 @@ export default function createEnableAlertTests({ getService }: FtrProviderContex
                 .auth(user.username, user.password)
                 .expect(200);
               expect(typeof updatedAlert.scheduled_task_id).to.eql('string');
+              expect(updatedAlert.execution_status.status).to.eql('pending');
```
Note to self: this is flaky in the scenario the rule finished running before the API call is made.

mikecote replied:
Had to remove the integration test: https://github.com/elastic/kibana/pull/98135/commits/cfb1c259ccacef9ccc08f4ac3578e9c328b76ece. It will always be a race condition between enabling the rule and checking if the execution status is `pending`. So we'll have to rely on unit tests.

mikecote commented on x-pack/test/alerting_api_integration/spaces_only/tests/alerting/enable.ts:
```diff
@@ -49,6 +49,7 @@ export default function createEnableAlertTests({ getService }: FtrProviderContex
         .set('kbn-xsrf', 'foo')
         .expect(200);
       expect(typeof updatedAlert.scheduled_task_id).to.eql('string');
+      expect(updatedAlert.execution_status.status).to.eql('pending');
```
Note to self: this is flaky in the scenario the rule finished running before the API call is made.

mikecote replied:
Had to remove the integration test: https://github.com/elastic/kibana/pull/98135/commits/cfb1c259ccacef9ccc08f4ac3578e9c328b76ece. It will always be a race condition between enabling the rule and checking if the execution status is `pending`. So we'll have to rely on unit tests.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/)
* Commit: 4c18b9f1c0e6ebd55ef686700fff56a7b8fa35e0
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-98135/4c18b9f1c0e6ebd55ef686700fff56a7b8fa35e0)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup12](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/execution/node/768/log)
* [Execute xpack-kibana-ciGroup12](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/execution/node/789/log)

### History
* :broken_heart: [Build #122160](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/) failed 6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b
* :broken_heart: [Build #122126](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/) failed c7b24857390ebeaec8cbcd53f92af5ff674cd553
* :green_heart: [Build #122027](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/) succeeded 7bca356f565f34d0410a7c0f90f1eb3978fc62fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @mikecote

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/","number":"122027","commit":"7bca356f565f34d0410a7c0f90f1eb3978fc62fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/","number":"122126","commit":"c7b24857390ebeaec8cbcd53f92af5ff674cd553"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/","number":"122160","commit":"6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/","number":"122775","commit":"4c18b9f1c0e6ebd55ef686700fff56a7b8fa35e0"}]}
PIPELINE-->

mikecote commented:
> >> re: ok -> recovered. I happened to think - "do we have doc to fix? maybe a screen shot?". Took a brief look - nope! Did I miss it? It ain't right that we don't have doc on the alert details UI! Let's open a new doc issue if we don't.
> >
> >https://www.elastic.co/guide/en/kibana/master/rule-details.html We do have a screenshot for this
> 
> I will update those for sure!

This is updated in https://github.com/elastic/kibana/pull/98135/commits/8b416440ef7dddc9fce35307429d03ba3f31f44b.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/123260/)
* Commit: cfb1c259ccacef9ccc08f4ac3578e9c328b76ece
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-98135/cfb1c259ccacef9ccc08f4ac3578e9c328b76ece)
* [Documentation Changes](https://kibana_98135.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b7719d65ee5f1f5689235fd6c99e17f0cabefdb3) | [after](https://github.com/elastic/kibana/commit/cfb1c259ccacef9ccc08f4ac3578e9c328b76ece) | diff |
| --- | --- | --- | --- |
| `triggersActionsUi` | 1.5MB | 1.5MB | +93.0B |

### History
* :broken_heart: [Build #122775](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/) failed 4c18b9f1c0e6ebd55ef686700fff56a7b8fa35e0
* :broken_heart: [Build #122160](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/) failed 6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b
* :broken_heart: [Build #122126](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/) failed c7b24857390ebeaec8cbcd53f92af5ff674cd553
* :green_heart: [Build #122027](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/) succeeded 7bca356f565f34d0410a7c0f90f1eb3978fc62fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @mikecote

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122027/","number":"122027","commit":"7bca356f565f34d0410a7c0f90f1eb3978fc62fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/","number":"122126","commit":"c7b24857390ebeaec8cbcd53f92af5ff674cd553"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/","number":"122160","commit":"6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/","number":"122775","commit":"4c18b9f1c0e6ebd55ef686700fff56a7b8fa35e0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/123260/","number":"123260","commit":"cfb1c259ccacef9ccc08f4ac3578e9c328b76ece"}]}
PIPELINE-->

mikecote commented:
@pmuellr @ymao1 this PR is good for another review 🙂

ymao1 reviewed (approved):
LGTM!

mikecote commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125798/)
* Commit: 07d5dfe06b465902f6b881cecdd127579bcdb6ff
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-98135/07d5dfe06b465902f6b881cecdd127579bcdb6ff)
* [Documentation Changes](https://kibana_98135.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/ee9d469295d80598ec1ce8142869987812fd9597) | [after](https://github.com/elastic/kibana/commit/07d5dfe06b465902f6b881cecdd127579bcdb6ff) | diff |
| --- | --- | --- | --- |
| `triggersActionsUi` | 1.5MB | 1.5MB | +93.0B |

<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/ee9d469295d80598ec1ce8142869987812fd9597) | [after](https://github.com/elastic/kibana/commit/07d5dfe06b465902f6b881cecdd127579bcdb6ff) | diff |
| --- | --- | --- | --- |
| `canvas` | 29 | 25 | -4 |
| `crossClusterReplication` | 8 | 6 | -2 |
| `fleet` | 4 | 2 | -2 |
| `globalSearch` | 4 | 2 | -2 |
| `indexManagement` | 12 | 7 | -5 |
| `infra` | 5 | 3 | -2 |
| `licensing` | 18 | 15 | -3 |
| `monitoring` | 109 | 56 | -53 |
| total |  |  | -73 |


</details>

### History
* :broken_heart: [Build #125282](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125282/) failed e57fc5ca7306c493f2f2de57122fae0d7333275d
* :green_heart: [Build #123260](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/123260/) succeeded cfb1c259ccacef9ccc08f4ac3578e9c328b76ece
* :broken_heart: [Build #122775](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/) failed 4c18b9f1c0e6ebd55ef686700fff56a7b8fa35e0
* :broken_heart: [Build #122160](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/) failed 6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b
* :broken_heart: [Build #122126](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/) failed c7b24857390ebeaec8cbcd53f92af5ff674cd553

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @mikecote

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122126/","number":"122126","commit":"c7b24857390ebeaec8cbcd53f92af5ff674cd553"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122160/","number":"122160","commit":"6b27795b7ba9b9d1f53d1012dfaa1346b0cf552b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122775/","number":"122775","commit":"4c18b9f1c0e6ebd55ef686700fff56a7b8fa35e0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/123260/","number":"123260","commit":"cfb1c259ccacef9ccc08f4ac3578e9c328b76ece"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125282/","number":"125282","commit":"e57fc5ca7306c493f2f2de57122fae0d7333275d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125798/","number":"125798","commit":"07d5dfe06b465902f6b881cecdd127579bcdb6ff"}]}
PIPELINE-->

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/100063)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/100063">](https://github.com/elastic/kibana/pull/100063) |

This backport PR will be merged automatically after passing CI.
